pkg/controller/job/job_controller.go

```diff
@@ -746,17 +746,19 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (forget bool, rEr
 	if feature.DefaultFeatureGate.Enabled(features.JobReadyPods) {
 		ready = pointer.Int32(countReadyPods(activePods))
 	}
-	// Job first start. Set StartTime and start the ActiveDeadlineSeconds timer
-	// only if the job is not in the suspended state.
+
+	// Job first start. Set StartTime only if the job is not in the suspended state.
 	if job.Status.StartTime == nil && !jobSuspended(&job) {
 		now := metav1.Now()
 		job.Status.StartTime = &now
-		// enqueue a sync to check if job past ActiveDeadlineSeconds
-		if job.Spec.ActiveDeadlineSeconds != nil {
-			klog.V(4).Infof("Job %s has ActiveDeadlineSeconds will sync after %d seconds",
-				key, *job.Spec.ActiveDeadlineSeconds)
-			jm.queue.AddAfter(key, time.Duration(*job.Spec.ActiveDeadlineSeconds)*time.Second)
-		}
+	}
+
+	// Let us queue the job back accounting to the activeDeadlineSeconds. The duration after which it should be
+	// re-queued is activeDeadlineSeconds - time spent after start time.
+	if !jobSuspended(&job) && job.Spec.ActiveDeadlineSeconds != nil {
+		syncDuration := time.Duration(*job.Spec.ActiveDeadlineSeconds) - time.Since(job.Status.StartTime.Time)
```

Comment:
@mimowo @alculquicondor 

```golang
	defer func() {
		// Let us queue the job back accounting to the activeDeadlineSeconds. The duration after which it should be
		// re-queued is activeDeadlineSeconds - time spent after start time.
		if !forget && !jobSuspended(&job) && job.Spec.ActiveDeadlineSeconds != nil {
			syncDuration := time.Duration(time.Duration(*job.Spec.ActiveDeadlineSeconds).Seconds()) - time.Duration(time.Since(job.Status.StartTime.Time).Seconds())
			klog.V(2).InfoS("Job has activeDeadlineSeconds configuration. Will sync this job again", "job", key, "duration", syncDuration, "status", job.Status)
			jm.queue.AddAfter(key, syncDuration)
		}
	}()
```
This seem to ensure that we do not re-queue it back after the item has been marked to be forgotten and addresses the time delta you were mentioning. Does this look all right ?

Reply:
cc @mimowo  @alculquicondor  after incorporating the changes suggested by @mimowo

Reply:
Hi, any progress on this?

Reply:
Pushing an update in a few hour.. Sorry. Got held up in some work related items.

Reply:
Changes pushed. PTAL.